### ISSUE TITLE
Using sigterm / catching sigterm to enable checkpointing

### ISSUE THREAD

**Issue by amueller** [awslabs/syne-tune#771]
Still new to this project, so I might be missing something.
From my understanding, for checkpointing, you're using ``checkpoint_model_at_rung_level(config, save_model_fn, epoch)`` to figure out whether to checkpoint at a given epoch or not.
I'm a bit surprised why this is necessary. Wouldn't it be possible to just checkpoint when the training process is stopped/paused by the scheduler?
Currently you're sending SIGKILL to the child processes, but I think you could also send SIGTERM, which could be caught to checkpoint the model before exiting, right?

Am I overlooking something?

**Comment by mseeger** [comment]
Hello Andreas,
we want to keep things as simple as possible on the side of the training script. In fact, in all our examples, checkpointing is done at the end of each epoch. Your suggestion would save time by checkpointing less often, but it would be more difficult to implement.

**Comment by amueller** [comment]
Ah ok, the documentation suggested using the helper functions, which seems more complicated than doing the exception as it requires passing the config objects around. But I saw that actually in the end it's always stored. So yes, it's more complicated than always storing but seems easier than the method that the code suggests was planned.

**Comment by amueller** [comment]
Thanks and thank you for all your replies! I think having the simple option of storing every epoch and the slightly more advanced, but probably also very simple to implement option of catching a sigterm for storing on demand would be great options to choose from for a user.
Basically, I got caught up in the docs that explain the helper functions and was a bit confused, also making clear that the recommended solution currently is to store every epoch also seems totally fine.

**Comment by geoalgo** [comment]
Btw, I forgot to mention we have this example that shows how to activate checkpointing for an xgboost script:
* example: https://github.com/awslabs/syne-tune/blob/main/examples/launch_checkpoint_example.py
* associated training script: https://github.com/awslabs/syne-tune/blob/main/examples/training_scripts/xgboost/xgboost_checkpoint.py

It may be useful as perhaps you are considering scikit-learn pipelines.

I am planning to have a look whether using sigterm works with checkpointing, I will let you know.

**Comment by amueller** [comment]
Thanks, I made checkpointing work, I'm actually using custom pytorch transformers 🤯
Crashed when the hard drive was full of checkpoints, now looking at your docs for fixing that :) The journey continues!

**Comment by mseeger** [comment]
https://syne-tune.readthedocs.io/en/latest/faq.html#checkpoints-are-filling-up-my-disk-what-can-i-do

The number of checkpoints stored only scales with the number of trials, not with how often a trial stores a checkpoint. Storing after every epoch simply just overwrites what was there before.

There is a speculative checkpoint removal feature, which is documented in the link above. Maybe this helps.

**Comment by jgolebiowski** [comment]
Hi @amueller, was that link helpful or do you need further support?

**Comment by amueller** [comment]
I'm good, thank you!

**Comment by jgolebiowski** [comment]
OK perfect, closing.
